### PR TITLE
Add debug assertions to write_bytes and copy*

### PR DESCRIPTION
**Pull request by RalfJung**
Looks like @nitnelave  went MIA in https://github.com/rust-lang/rust/pull/58783, so I am re-submitting their PR, tweaked just a bit. I took care to preserve commit authorship.

Cc https://github.com/rust-lang/rust/issues/53871

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -1331,6 +1333,26 @@ extern "rust-intrinsic" {
 // (`transmute` also falls into this category, but it cannot be wrapped due to the
 // check that `T` and `U` have the same size.)
 
+/// Checks whether `ptr` is properly aligned with respect to
+/// `align_of::<T>()`.
+pub(crate) fn is_aligned_and_not_null<T>(ptr: *const T) -> bool {
+    return !ptr.is_null() && ptr as usize % mem::align_of::<T>() == 0;
```
Comment (Centril) [inline src/libcore/intrinsics.rs]:
```suggestion
    !ptr.is_null() && ptr as usize % mem::align_of::<T>() == 0
```

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -1331,6 +1333,26 @@ extern "rust-intrinsic" {
...
+}
+
+/// Checks whether the regions of memory starting at `src` and `dst` of size
+/// `count * size_of::<T>()` overlap.
+fn overlaps<T>(src: *const T, dst: *const T, count: usize) -> bool {
+    let src_usize = src as usize;
+    let dst_usize = dst as usize;
+    let size = mem::size_of::<T>().checked_mul(count).unwrap();
+    let diff = if src_usize > dst_usize {
```
Comment (Centril) [inline src/libcore/intrinsics.rs]: I'm surprised there's no utility function in libcore for this... should there be one?

File: src/librustc_codegen_llvm/metadata.rs [hunk 3]
```diff
@@ -67,10 +66,16 @@ fn search_meta_section<'a>(of: &'a ObjectFile,
     unsafe {
         let si = mk_section_iter(of.llof);
         while llvm::LLVMIsSectionIteratorAtEnd(of.llof, si.llsi) == False {
-            let mut name_buf = ptr::null();
+            let mut name_buf = None;
             let name_len = llvm::LLVMRustGetSectionName(si.llsi, &mut name_buf);
-            let name = slice::from_raw_parts(name_buf as *const u8, name_len as usize).to_vec();
-            let name = String::from_utf8(name).unwrap();
+            let name = name_buf.map_or(
+                String::new(), // we got a NULL ptr, ignore `name_len`
```
Comment (Centril) [inline src/librustc_codegen_llvm/metadata.rs]:
```suggestion
                String::new(), // We got a NULL ptr, ignore `name_len`.
```

**Comment by alexcrichton** [comment]
Seems reasonable to me! r=me with @Centril's nits

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/intrinsics.rs]: For what? The absolute difference of two unsigned integers?

I don't know if there should be one. ;) Where would it even be?

**Comment by RalfJung** [comment]
@alexcrichton I had to disable some codegen tests in debug mode, because even with optimizations the new debug assertions get in the way.

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (Centril) [reply src/libcore/intrinsics.rs]: Yeah the absolute diff... Somewhere in `core::num` or as methods on the types?

Comment (RalfJung) [reply src/libcore/intrinsics.rs]: Hm, not sure if this PR is the right place to introduce some new unstable public API?

Comment (Centril) [reply src/libcore/intrinsics.rs]: Yeah I agree, it was an aside remark, sorry should have been more clear :)

Comment (Centril) [reply src/libcore/intrinsics.rs]: Filed as https://github.com/rust-lang/rust/issues/62111.

Comment (RalfJung) [reply src/libcore/intrinsics.rs]: Thanks!

**Comment by RalfJung** [comment]
So the nits are fixed, but I'd like to get explicit confirmation that it is okay to ignore some (more) codegen tests in debug mode. We do run some CI runners in release mode, right?
(These are not the first codegen tests to be disabled in debug mode.)

**Comment by Mark-Simulacrum** [comment]
@bors r-

This fails in local testing after rebasing onto master with from libcore doctests. 

```
---- ptr/mod.rs - ptr::drop_in_place (line 158) stdout ----
Test executable failed (exit code 101).

stderr:
thread 'main' panicked at 'attempt to copy from unaligned or null pointer', src/libcore/intrinsics.rs:1510:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.



failures:
    ptr/mod.rs - ptr::drop_in_place (line 158)
```

**Comment by RalfJung** [comment]
Oh... actually that's nice! We just rediscovered https://github.com/rust-lang/rust/issues/62313. The assertions are doing something. ;)

**Comment by RalfJung** [comment]
Blocked on https://github.com/rust-lang/rust/pull/62351.

**Comment by lzutao** [comment]
#62351 is merged 
@rustbot modify labels: - S-blocked +S-waiting-on-author

**Comment by mati865** [comment]
This PR failed in #62372 and #62378 with error:
```
error: linking with `arm-linux-gnueabihf-gcc` failed: exit code: 1
    |
    = note: "arm-linux-gnueabihf-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/debuginfo-lto/debuginfo-lto.debuginfo_lto.7rcbfp3g-cgu.1.rcgu.o" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/debuginfo-lto/debuginfo-lto" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/checkout/obj/build/arm-unknown-linux-gnueabihf/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/debuginfo-lto/auxiliary" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/tmp/rustcwZ9vAg/libbacktrace_sys-69828ba5b177f6e2.rlib" "-Wl,--end-group" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib/libcompiler_builtins-34ad96caca8cd606.rlib" "-Wl,-Bdynamic" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil" "-Wl,-rpath,$ORIGIN/auxiliary" "-Wl,-rpath,$ORIGIN/../../../stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "-Wl,-rpath,/checkout/obj/lib/rustlib/arm-unknown-linux-gnueabihf/lib" "-Wl,--enable-new-dtags"
    = note: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-unknown-linux-gnueabihf/lib/libcompiler_builtins-34ad96caca8cd606.rlib(compiler_builtins-34ad96caca8cd606.compiler_builtins.7arsbjs9-cgu.0.rcgu.o): In function `__aeabi_memcpy8':
                  compiler_builtins.7arsbjs9-cgu.0:(.text.__aeabi_memcpy4+0x6c): undefined reference to `core::panicking::panic::h747b6b606f8c4f43'
                 compiler_builtins.7arsbjs9-cgu.0:(.text.__aeabi_memcpy4+0x7c): undefined reference to `core::panicking::panic::h747b6b606f8c4f43' collect2: error: ld returned 1 exit status 
```

**Comment by RalfJung** [comment]
Yeah looks like this one s at fault there.

@bors r-

However, I don't have the slightest idea why. Seems like Android does not support panics, or so? Who can one Cc for weird android linker issues?

**Comment by bjorn3** [comment]
I believe it is expected that it cant find panic: the library needing a symbol must be before the definition of the symbol. As `libcompiler_builtins` is the last rust library, it cant find panic. I think the question is why `libcore` is build with debug assertions on android.

**Comment by mati865** [comment]
The error seems to come from `compiler-builtins` and maintainer of that repo is already assigned to this PR.

**Comment by RalfJung** [comment]
Building libcore with debug assertions works just fine though locally for this PR.

**Comment by mati865** [comment]
`__aeabi_memcpy4` is only defined on arm and calls https://github.com/rust-lang-nursery/compiler-builtins/blob/39ad0538f2cfbc81b62e4f0ea1ee7a2f95133094/src/arm.rs#L144
Then `ptr::read` calls https://github.com/rust-lang/rust/blob/b820c761744db080ff7a4ba3ac88d259065cb836/src/libcore/ptr/mod.rs#L601 where you have added assertions.

This together with bjron3 comment should explain it.

**Comment by bjorn3** [comment]
I think adding a second `-llibcompiler_builtins-*.rlib` at the beginning of the linker invocation could fix this.

**Comment by RalfJung** [comment]
> __aeabi_memcpy4 is only defined on arm and calls ptr::read

I see, that makes sense.

> I think adding a second -llibcompiler_builtins-*.rlib at the beginning of the linker invocation could fix this.

I am afraid I have no idea how linkers are invoked in general and in Rust specifically. My current inclination is to add some `cfg` so that `ptr::read` only does these checks on x86, but that would be sad. Since you seem to have a pretty clear idea for what it takes, could you open a PR against https://github.com/RalfJung/rust/tree/debug-assert ?

**Comment by nikic** [comment]
@RalfJung Replacing `ptr::write(dest, ptr::read(src))` with `*dest = *src` would probably do it as well.

**Comment by RalfJung** [comment]
So is there some kind of guarantee that the arguments of `__aeabi_memcpy4` are 4-aligned?

**Comment by nikic** [comment]
@RalfJung Yes, that's exactly the difference between `memcpy` and `memcpy4`, the latter has guaranteed 4-byte alignment.

**Comment by alexcrichton** [comment]
I would personally prefer to avoid updating any linkage with compiler-builtins, in general compiler-builtins should never be referencing panicking symbols. I agree with @nikic that the best solution for this is probably source-level changes in the compiler-builtins crate to avoid these assertions getting embedded, even in release mode.

**Comment by RalfJung** [comment]
All right, submitted PR for compiler-builtins: https://github.com/rust-lang-nursery/compiler-builtins/pull/300.

**Comment by RalfJung** [comment]
I rebased the PR and added a commit that bumps compiler_builtins to the fresh 0.1.17 release.

**Pull request merged**